[Feature] [Plugin] Automatyczne wysyłanie wiadomości PW do nowo zarejestrowanego użytkownika

Mariusz08 opened the pull request:
Tak jak w tytule; plugin ma za zadanie umożliwić userowi zapoznać się z systemem wiadomości prywatnych. Sama wiadomość jest ustalana w panelu administratora a także ID bota (usera od którego będzie szła wiadomość)

Ps. Wie ktoś może jak wysłać powiadomienie do usera? Trzeba chyba dodać jakiś wpis w  qa_eventlog ale dodałem ich ok. 70 a żaden nie działa. Jakieś pomysły? ;)

awaluk commented:
Co do kodu chyba nie mam zastrzeżeń, wiele tu nie ma :) Nie bardzo widzę sens oznaczania wielkimi czerwonymi literami pola z id użytkownika, ale ok. Dodałbym przed wysłaniem sprawdzenie czy treść wiadomości nie jest pusta - jeśli nawet jest włączone wysyłanie a treść nie jest wpisana to raczej nie ma sensu. Wydaje się też, że należałoby sprawdzać czy user o podanym id istnieje, bo jeśli nie to co? Wiadomość przyjdzie jako "od nikogo" albo wcale się nie wyśle? Ponadto w ustawieniach pluginu nie da się już wyłączyć wysyłania - jest włączone, odznaczam checbkoxa, biorę zapisz i nadal jest włączone.

Samo wysłanie wiadomości działa. Wysyłana jest jednak tylko wiadomość na forum, użytkownik nie dostaje jej emailem - czy to dobrze czy źle, to sam nie wiem. Może dobrze, bo od razu po rejestracji i tak dostaje maila z powitaniem i linkiem do aktywacji konta. Źle bo można się pogubić jak to PW w końcu działa - czy wysyła informacje o nich mailem czy nie (a domyślnie wszystkie wysyła). Tak więc w sumie sam nie wiem.

No i powiadomienie jak napisałeś nie działa, a to jest raczej kluczowy element całości, bo tak to nikt się o tej wiadomości nie dowie. Przeanalizowałbym jak one są wyciągane do wyświetlenia, to chyba [tutaj](https://github.com/CodersCommunity/forum.pasja-informatyki.local/blob/master/forum/qa-plugin/q2apro-on-site-notifications/q2apro-onsitenotifications-page.php#L75), może to coś pomoże.

Jeśli miałoby to wejść to chętnie pozbieram też sugestie jak chcemy nazwać konto wysyłające (czy po prostu "Forum Pasja Informatyki" lub "Forum" czy może coś bardziej wymyślnego) oraz jaka ma być finalnie treść wiadomości - żeby nie było ani za długo ani za krótko, już dyskusja była w #122 ale w sumie nic z niej nie wyniknęło.

Mariusz08 commented:
> Co do kodu chyba nie mam zastrzeżeń, wiele tu nie ma :)

Mój pierwszy PR bez zastrzeżeń do kodu :wink:

>  Nie bardzo widzę sens oznaczania wielkimi czerwonymi literami pola z id użytkownika, ale ok.

Tak jak jest napisane - `DANGER ZONE!` - strefa niebezpieczna, takie ostrzeżenie, żeby się dobrze zastanowić przed zmianą tego pola bo trochę słabo, gdy nagle wiadomości będą przechodzić np. ode mnie ;)

> Dodałbym przed wysłaniem sprawdzenie czy treść wiadomości nie jest pusta - jeśli nawet jest włączone wysyłanie a treść nie jest wpisana to raczej nie ma sensu.

W sumie faktycznie, nie pomyślałem, jutro wrzucę poprawkę.

> Wydaje się też, że należałoby sprawdzać czy user o podanym id istnieje, bo jeśli nie to co? Wiadomość przyjdzie jako "od nikogo" albo wcale się nie wyśle?

Z tego co pamiętam w takiej sytuacji wiadomość zostanie wysłana od użytkownika `niezalogowany`, aczkolwiek też wrzucę jutro poprawkę.

> Ponadto w ustawieniach pluginu nie da się już wyłączyć wysyłania - jest włączone, odznaczam checbkoxa, biorę zapisz i nadal jest włączone.

To dziwne, wcześniej u mnie działało. Przyjrzę się temu.

> Jeśli miałoby to wejść to chętnie pozbieram też sugestie jak chcemy nazwać konto wysyłające (czy po prostu "Forum Pasja Informatyki" lub "Forum" czy może coś bardziej wymyślnego)

Wg mnie prosto - `Forum Pasja Informatyki`, `Pasja Informatyki` albo `Bot Pasja Informatyki` - może kiedyś będziemy chcieli wysyłać jeszcze jakieś masowe wiadomości do userów

>  jaka ma być finalnie treść wiadomości - żeby nie było ani za długo ani za krótko, już dyskusja była w #122 ale w sumie nic z niej nie wyniknęło.

Treść jest raczej do przedyskutowania - czy wiadomość bardziej krótka, żeby zademonstrować działanie PW, czy może trochę dłuższa, ale przekazująca ważne informacje?

awaluk commented:
Obawiam się, że dłuższą mało kto przeczyta, dlatego raczej stawiałbym na nieco krótszą niż prezentowany w #122 przez Ciebie przykład, ale też może i nieco dłuższą niż podana przez @PanLydka :P Do ustalenia jeszcze w trakcie, przede wszystkim musi zadziałać powiadomienie.

JSHoliq commented:
Ja bym nadawał krótką wiadomość powitalną, ponieważ na Forum są już podane informacje dotyczące zasad korzystania i inne metadane: regulamin, FAQ, manifest, porady nie od parady, ekipa ninja, zasady przyznawania punktów. Dlatego w propozycji podanej w #122 wspomniał bym jeszcze o charakterze udzielanej pomocy i tyle:

"Witaj na Forum Pasja Informatyki! Podstawowe odpowiedzi na pytania jak korzystać z forum znajdziesz w naszym dziale FAQ. Pamiętaj, że na forum skupiamy się głównie na pomocy w rozwiązaniu problemów programistycznych, a nie udzielaniu gotowych odpowiedzi na pytania. Powodzenia! :)"

Na nazwę konta wysyłającego wg mnie odpowiednie będzie "Forum Pasja Informatyki", gdyż nazwa ta wskazuje na... nazwę forum, po prostu. :)

Mariusz08 commented:
Udało mi się znaleźć sposób na wysyłanie powiadomienia dla użytkownika. Plugin działa :)

awaluk commented:
> Dodałbym przed wysłaniem sprawdzenie czy treść wiadomości nie jest pusta - jeśli nawet jest włączone wysyłanie a treść nie jest wpisana to raczej nie ma sensu. 

To nadal zachowuje się tak samo, wysyła mi puste wiadomości.

> Wydaje się też, że należałoby sprawdzać czy user o podanym id istnieje, bo jeśli nie to co? 

To zaraz po rejestracji wyrzuca że wystąpił jakiś błąd Q2A, czyli sprawdzania nadal nie ma. Może weryfikacja poprawnego id usera i nie pustej treści wiadomości powinna być już w momencie włączania pluginu w panelu - tylko w momencie poprawnego wypełnienia konfiguracji plugin zostanie włączony?

> Ponadto w ustawieniach pluginu nie da się już wyłączyć wysyłania - jest włączone, odznaczam checbkoxa, biorę zapisz i nadal jest włączone.

Tu też bez zmian.

Powiadomienie przychodzi, fajnie. Ale coś jest z nim nie tak, bo gdy próbuję otworzyć listę powiadomień (klikam na "1") to kończy się tak:
![zrzut ekranu z 2018-07-28 18-49-43](https://user-images.githubusercontent.com/13801608/43358761-042a4a5a-9297-11e8-9f65-cffa685b53ed.png)

Mariusz08 commented:
> To nadal zachowuje się tak samo, wysyła mi puste wiadomości.

fixed
> To zaraz po rejestracji wyrzuca że wystąpił jakiś błąd Q2A, czyli sprawdzania nadal nie ma. Może weryfikacja poprawnego id usera i nie pustej treści wiadomości powinna być już w momencie włączania pluginu w panelu - tylko w momencie poprawnego wypełnienia konfiguracji plugin zostanie włączony?

fixed
> Tu też bez zmian.

fixed

> Powiadomienie przychodzi, fajnie. Ale coś jest z nim nie tak, bo gdy próbuję otworzyć listę powiadomień (klikam na "1") to kończy się tak:

fixed

Wszystko  naprawione, powinno już działać, jak są jakieś błędy to dajcie znać.

awaluk commented:
W działaniu od strony użytkownika jest ok, ale w panelu administracyjnym nie dostaję żadnej informacji o błędzie - gdy np. tylko włączę plugin i nie wpiszę treści ani id usera, to strona ustawień się jakby odświeża i nic. Ani się to nie zapisuje, ani nie ma błędu z wyjaśnieniem co się stało. Nadal nie mogę żadnym sposobem wyłączyć pluginu, czy tylko odznaczę opcję włączenia, czy też usunę wartości wszystkich pól, po zapisaniu strona się przeładowuje i wartości nadal są, plugin jest włączony. Podobnie gdy wpiszę id nieistniejącego użytkownika nie ma żadnego błędu, a strona się po prostu przeładowuje.

Mariusz08 commented:
Solved.

awaluk commented:
Może to już będzie że się czepiam, ale jak widzę zielony komunikat to jestem przekonany, że to jest sukces, a tam "invalid bot id". Mylące, dałbym standardowo czerwony, jak mamy wszędzie indziej.

Poza tym gdy próbuję zapisać ustawienia, a nie zaznaczę "enable plugin" to nie zapisuje się nic. Niby nie jest to wielki problem, ale to znowu jest mylące. Załóżmy, że przygotuję sobie treść, którą chcę wysyłać, ale jeszcze nie chcę włączyć, więc wpisuje, klikam zapisz i nic, nie zapisuje się, nie dostaję żadnego błędu. Rozumiałbym walidację w stylu gdy zaznaczę checkboxa, a nie wypełniłem treści albo id, to miałoby sens, ale wtedy też wypadałoby dać jakiś komunikat. Jak już z kolei włączę, później wyłączę i chcę edytować treść na wyłączonym to znów się nie da - klikam zapisz i nic się nie zapisuje dopóki nie włączę.

Ja wiem że to tylko panel dla administratora, zmienione to będzie pewnie sporadycznie i jak teraz to dodam to będzie ok, ale jak wrócę za rok to się będę zastanawiał o co tu chodzi, że to się nie zapisuje. Niech to będzie jakoś minimalnie intuicyjne.

event15 commented on forum/qa-plugin/send-pw-on-register/metadata.json:
```diff
@@ -0,0 +1,13 @@
+    "update_uri": "https://github.com/CodersCommunity/forum.pasja-informatyki.local",
+    "min_q2a": "1.7",
+    "min_php": "5.3"
+}
```
brak nowej lini na końcu pliku

event15 commented on forum/qa-plugin/send-pw-on-register/qa-plugin.php:
```diff
@@ -0,0 +1,34 @@
+  'qa_pw_event',
+  'PW Event');
+
+?>
```
nie powinno być `?>` oraz brak nowej linii na końcu pliku

awaluk commented:
> Może to już będzie że się czepiam, ale jak widzę zielony komunikat to jestem przekonany, że to jest sukces, a tam "invalid bot id". Mylące, dałbym standardowo czerwony, jak mamy wszędzie indziej.

Bez zmian

Gdy wpiszę id bota i treść, a nie włączę plugina, to zapisuje się treść, id bota nie zapisuje się wcale bez żadnego komunikatu.

Mam włączony plugin, wpisane id bota i treść. Wchodzę, zmieniam id na błędne, zapisuje. Dostaje komunikat (na zielono czyli nadal myli), że id jest błędne, więc się nie zapisuje i zostaje stare id, ale jednocześnie wyłącza plugin.

Podtrzymuję to co napisałem komentarz wcześniej. Postaw się w roli użytkownika, jakbyś to Ty dostał ten plugin i korzystał z niego pierwszy raz:
> Ja wiem że to tylko panel dla administratora, zmienione to będzie pewnie sporadycznie i jak teraz to dodam to będzie ok, ale jak wrócę za rok to się będę zastanawiał o co tu chodzi, że to się nie zapisuje. Niech to będzie jakoś minimalnie intuicyjne.

Samo wysyłanie PW jest w porządku. Ale przyszło mi do głowy sprawdzenie jeszcze jednej rzeczy. Włączyłem plugin, wpisałem w ustawieniach id usera, którego następnie usunąłem i spróbowałem się zarejestrować. Prawdopodobnie taki przypadek u nas nie wystąpi, jednak wydaje się, że powinno to być zabezpieczone. Efekt to niestety niedziałająca rejestracja:
> ( ! ) Notice: Undefined variable: messageId in /vagrant/forum/qa-plugin/send-pw-on-register/qa-pw-event.php on line 31
A Question2Answer database query failed when generating this page.
A full description of the failure is available in the web server's error log file.

event15 commented on forum/qa-plugin/send-pw-on-register/qa-pw-event.php:
```diff
@@ -7,11 +7,12 @@ public function process_event($event, $userId, $handle, $cookieId, $params)
             $messageContent = qa_opt('sendPwMessageOnRegister_messageContent');
             
-            if (empty($botId) || empty($messageContent)) {
+            if (empty($botId) || empty($messageContent) || [] === qa_db_user_get_userid_handles(qa_post_text('botId'))) {
```
co to za dziwny twór?
